### PR TITLE
[IMP] mass_mailing: remove stats computation

### PR DESCRIPTION
**Pull request by datonefaridze**
In order to make kanban loading faster we removed
computation of various statistics.

Additionally, tests were added to make sure that future 
changes to statistics computation is valid.

Also, the performance of computation of statistics is improved by
adding indexes and modifying compute function. Some tests are
also added to make sure the reliability of modified function.

task-3485424

### CODE REVIEW

File: addons/mass_mailing/views/mailing_mailing_views.xml [hunk 1]
```diff
@@ -539,31 +532,13 @@
...
-                                            <div name="stat_replied">
-                                                <b><field name="replied_ratio" />%</b> Replied
-                                            </div>
-                                            <div name="stat_clicks">
-                                                <b><field name="clicks_ratio" />%</b> Clicks
-                                            </div>
-                                        </div>
-                                    </div>
                                 </div>
+                                <i class="fa fa-bullseye" role="img" aria-label="Lead/Opportunity"> Lead/Opportunity</i> <br />
```
Comment (jeh-odoo) [inline addons/mass_mailing/views/mailing_mailing_views.xml:536]: Suggestion to keep the correct font and add a margin to align this part with the line below
```suggestion
                                <i class="fa fa-bullseye me-2" role="img" aria-label="Lead/Opportunity"/> Lead/Opportunity
```

**Review by jeh-odoo — COMMENTED** [review]
Hello @datonefaridze 
Just a small suggestion to keep the correct font and alignment.
Thanks for the changes!
Cheers!

**Comment by datonefaridze** [comment]
@jeh-odoo done

File: addons/mass_mailing/views/mailing_mailing_views.xml [hunk 2]
```diff
@@ -539,31 +532,15 @@
...
-                                                <b><field name="replied_ratio" />%</b> Replied
-                                            </div>
-                                            <div name="stat_clicks">
-                                                <b><field name="clicks_ratio" />%</b> Clicks
-                                            </div>
-                                        </div>
-                                    </div>
                                 </div>
+                                <i class="fa fa-bullseye me-2" role="img" aria-label="Lead/Opportunity"/>
+                                <field name='mailing_model_id' invisible="mailing_on_mailing_list"/>
```
Comment (std-odoo) [inline addons/mass_mailing/views/mailing_mailing_views.xml:537]:
```suggestion
                                <field name="mailing_model_id" invisible="mailing_on_mailing_list"/>

**Review by std-odoo — COMMENTED** [review]
@datonefaridze Hi :) 

Only one picky comment, otherwise LGTM :) 

I tried to make `_compute_statistics` and `_compute_clicks_ratio` faster, but just that query is slow, so not sure it's possible to make it faster...
```sql
    SELECT m.id as mailing_id,
           COUNT(s) AS expected
      FROM mailing_mailing m
 LEFT JOIN mailing_trace s
        ON s.mass_mailing_id = m.id
  GROUP BY m.id;
```

File: addons/mass_mailing/views/mailing_mailing_views.xml (hunk 2 above)
Comment (datonefaridze) [reply addons/mass_mailing/views/mailing_mailing_views.xml:537]: @std-odoo currently we aren't focused on speed (as far as i know), we might do it in future.

**Comment by std-odoo** [comment]
@datonefaridze Re :) 

With the precious help of @ryv-odoo , we managed to improve a lot the perf of the 2 problematic compute (`_compute_statistics` and `_compute_clicks_ratio`)

Before
image

After
image

Here is the branch: https://github.com/odoo-dev/odoo/tree/master-amss_mailing-perf-2023-std

I think we can try to do it in this PR, in a separate commit (but keep the change in the kanban view)

Cheers

**Comment by datonefaridze** [comment]
@std-odoo cool, improvement is huge. I will take a look later today. Thanks

File: addons/mass_mailing/models/mailing.py [hunk 3]
```diff
@@ -304,43 +305,39 @@ def _compute_statistics(self):
...
-            row['bounced_ratio'] = float_round(100.0 * row['bounced'] / total, precision_digits=2)
-            self.browse(row.pop('mailing_id')).update(row)
+
+        result = self.env["mailing.trace"].sudo()._read_group(
+            [("mass_mailing_id", "in", self.ids)],
+            ['mass_mailing_id', 'trace_status'],
+            ['__count', 'links_click_datetime:count', 'sent_datetime:count'])
+
+        result_per_mailing = defaultdict(lambda: defaultdict(int))
+        for line in result:
```
Comment (std-odoo) [inline addons/mass_mailing/models/mailing.py:315]: you can deconstruct line like that
```suggestion
        for mailing, count, links_click_datetime, sent_datetime in result:

File: addons/test_mass_mailing/tests/test_mailing_statistics_sms.py [hunk 4]
```diff
@@ -43,14 +43,24 @@ def test_mailing_statistics_sms(self):
...
 
-        # check mailing statistics
         self.assertEqual(mailing.clicked, 3)
         self.assertEqual(mailing.delivered, 4)
-        self.assertEqual(mailing.received_ratio, 40)
         self.assertEqual(mailing.opened, 3)
-        self.assertEqual(mailing.opened_ratio, 30)
         self.assertEqual(mailing.sent, 10)
 
+        self.assertEqual(mailing.scheduled, 0)  # TODO: not zero values here
```
Comment (std-odoo) [inline addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:51]: can you make this non zero ? (with a commit before that one, to be sure it doesn't change with the fix)

File: addons/test_mass_mailing/tests/test_mailing_statistics_sms.py [hunk 5]
```diff
@@ -43,14 +43,24 @@ def test_mailing_statistics_sms(self):
...
+        self.assertEqual(mailing.process, 0)
+        self.assertEqual(mailing.pending, 6)
+        self.assertEqual(mailing.replied, 0)
+        self.assertEqual(mailing.bounced, 0)
+        self.assertEqual(mailing.failed, 0)
+
+        self.assertEqual(mailing.received_ratio, 40)
+        self.assertEqual(mailing.opened_ratio, 30)
+        self.assertEqual(mailing.replied_ratio, 0)
+        self.assertEqual(mailing.bounced_ratio, 0)
```
Comment (std-odoo) [inline addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:62]: same here

**Review by std-odoo — COMMENTED** [review]
@datonefaridze Bloup :)

File: addons/test_mass_mailing/tests/test_mailing_statistics_sms.py (hunk 4 above)
Comment (datonefaridze) [reply addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:51]: @std-odoo how many of these do you want to be non zero?
```
        self.assertEqual(mailing.scheduled, 0)  # TODO: not zero values here
        self.assertEqual(mailing.canceled, 0)
        self.assertEqual(mailing.process, 0)
        self.assertEqual(mailing.pending, 6)
        self.assertEqual(mailing.replied, 0)
        self.assertEqual(mailing.bounced, 0)
        self.assertEqual(mailing.failed, 0)
        ```
       all of them or just `mailing.scheduled` ?

Comment (std-odoo) [reply addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:51]: all

Comment (datonefaridze) [reply addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:51]: @std-odoo i don't think we can do all. when testing sms we don't have reply flow created, we don't even detect it. same goes to many other flows: failed, process (those flows are for mailings). i will think how i can do scheduled, canceled and bounced. those

Comment (std-odoo) [reply addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:51]: can't you manually create the traces ?

Comment (datonefaridze) [reply addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:51]: yes but it won't make sense, we are testing mailing feature on sms @std-odoo

Comment (std-odoo) [reply addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:51]: ah yeah indeed, we should do an other test so, sorry (so not in mass mailing sms, but just in mass mailing)

Comment (datonefaridze) [reply addons/test_mass_mailing/tests/test_mailing_statistics_sms.py]: @std-odoo done, i removed replied because it is already being tested in mail test. I only left `scheduled` and `process` because function doesn't allow me to capture those states, send function just does it right away and writes `sent` state at  the end

Comment (std-odoo) [reply addons/test_mass_mailing/tests/test_mailing_statistics_sms.py]: @datonefaridze Can you do it in a different commit ? Before other commits (that way we can run the test before and after your changes)

**Comment by datonefaridze** [comment]
@std-odoo can we also merge this one?

File: addons/test_mass_mailing/tests/test_mailing_statistics_sms.py [hunk 6]
```diff
@@ -43,13 +68,36 @@ def test_mailing_statistics_sms(self):
         for record_idx in (0, 2, 3):
             self.gateway_sms_click(mailing, target_records[record_idx])
 
-        # check mailing statistics
+        for record_idx in [4, 5, 6]:
```
Comment (std-odoo) [inline addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:71]: this is still tested on a sms mailing no ?

image

**Review by std-odoo — COMMENTED** [review]
@datonefaridze Hi

About the test I tough we would do it on a "email" mailing, not on a "sms" mailing (as we discussed in the previus review)

File: addons/test_mass_mailing/tests/test_mailing_statistics_sms.py [hunk 7]
```diff
@@ -8,7 +8,7 @@
 from odoo.tests.common import users
 from odoo.tests import tagged
 from odoo.tools import mute_logger
-
+from odoo.addons.phone_validation.tools import phone_validation
```
[the remark this answers is not in the document]
Comment (datonefaridze) [reply addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:11]: @std-odoo its not 3 anymore (first one came from master, now i rebased and it's 2 again).

[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:11]: yeah so improve the PR message to reflect the *2* commits ^^

[the remark this answers is not in the document]
Comment (datonefaridze) [reply addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:11]: @std-odoo yes
image

File: addons/test_mass_mailing/tests/test_mailing_statistics_sms.py [hunk 8]
```diff
@@ -43,13 +68,28 @@ def test_mailing_statistics_sms(self):
...
-        # check mailing statistics
+        for record_idx in [7, 8]:
+            record = target_records[record_idx]
+            trace = mailing.mailing_trace_ids.filtered(lambda t: t.model == record._name and t.res_id == record.id)
+            trace.set_bounced()
+
+        for record_idx in [9, 10, 11, 12]:
+            record = target_records[record_idx]
+            trace = mailing.mailing_trace_ids.filtered(lambda t: t.model == record._name and t.res_id == record.id)
+            trace.set_canceled()
```
Comment (tde-banana-odoo) [inline addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:79]: Hello, would be better to have real canceled traces, due to invalid / duplicate number for example. always better to test a functional flow if possible

File: addons/test_mass_mailing/tests/test_mailing_statistics_sms.py [hunk 9]
```diff
@@ -25,6 +25,31 @@ def setUpClass(cls):
             signature='--\nMarie'
         )
 
+        records = cls.records
+        partners = cls.partners
+        country_be_id = cls.env.ref('base.be').id
+        _country_us_id = cls.env.ref('base.us').id
+
+        for x in range(10):
```
Comment (tde-banana-odoo) [inline addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:33]: Looks globally like a duplicate of TestMassSMSCommon init, are you adding new records for test purpose ?

File: addons/test_mass_mailing/tests/test_mailing_statistics_sms.py [hunk 10]
```diff
@@ -25,6 +25,31 @@ def setUpClass(cls):
...
+            partners += cls.env['res.partner'].with_context(**cls._test_context).create({
+                'name': 'Partner_1_%s' % (x),
+                'email': '_test_partner_1_%s@example.com' % (x),
+                'country_id': country_be_id,
+                'mobile': '045300%s%s99' % (x, x)
+            })
+            records += cls.env['mail.test.sms'].with_context(**cls._test_context).create({
+                'name': 'MassSMSTest_1_%s' % (x),
+                'customer_id': partners[x].id,
+                'phone_nbr': '045300%s%s44' % (x, x)
```
Comment (tde-banana-odoo) [inline addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:43]: all that could be moved in a small helper (a bit like '_create_mailing_test_records')

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, saw some light, added some comments :)

File: addons/test_mass_mailing/tests/test_mailing_statistics_sms.py (hunk 9 above)
Comment (datonefaridze) [reply addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:33]: @tde-banana-odoo yes, is that wrong? i needed more examples to create failed/canceled/clicked and  other traces

**Comment by datonefaridze** [comment]
@tde-banana-odoo can we merge this one?

File: addons/test_mass_mailing/tests/test_mailing_statistics.py [hunk 11]
```diff
@@ -53,6 +53,7 @@ def test_mailing_statistics(self):
         self.assertEqual(mailing.bounced, 1)
         self.assertEqual(mailing.bounced_ratio, 8.33)
         self.assertEqual(mailing.canceled, 1)
+        self.assertEqual(mailing.expected, 13)
```
Comment (std-odoo) [inline addons/test_mass_mailing/tests/test_mailing_statistics.py:56]: this change should be in the previous commit no ?

Comment (tde-banana-odoo) [reply addons/test_mass_mailing/tests/test_mailing_statistics.py:56]: @datonefaridze please read and update all comments, you mark as solved but it is not done. Don't try to be faster than light, in the end we lose time having to recheck the PR content to be sure everything is done as intended.

File: addons/test_mass_mailing/tests/test_mailing_statistics_sms.py [hunk 12]
```diff
@@ -93,7 +93,7 @@ def test_mailing_statistics_sms(self):
         kpi_values = body_html.xpath('//table[@data-field="sms"]//*[hasclass("kpi_value")]/text()')
         self.assertEqual(
             [t.strip().strip('%') for t in kpi_values],
-            ['25.0', str(float(mailing.clicks_ratio)), str(float(mailing.bounced_ratio))]
+            ['25.0', str(float(mailing.opened_ratio)), str(float(mailing.bounced_ratio))]
```
Comment (tde-banana-odoo) [inline addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:96]: This seems unrelated (and wasn't it fixed in stable anyway ?)

File: addons/test_mass_mailing/tests/test_mailing_statistics.py (hunk 11 above)
Comment (datonefaridze) [reply addons/test_mass_mailing/tests/test_mailing_statistics.py:56]: @tde-banana-odoo i also checked this comment (asked my self the same question) and i thought std meant this comment (because first commit is about ui).

Comment (datonefaridze) [reply addons/test_mass_mailing/tests/test_mailing_statistics.py:56]: @tde-banana-odoo Now i think i did everything correctly. (checked everything too, hope i am not making any mistakes)

File: addons/test_mass_mailing/tests/test_mailing_statistics_sms.py [hunk 13]
```diff
@@ -24,6 +24,22 @@ def setUpClass(cls):
...
             signature='--\nMarie'
         )
+        records = cls.records
+        partners = cls.partners
+
+        mobile_numbers = ['045300%s%s99' % (x, x) for x in range(6)]
+        mobile_numbers += ['0453000099', '0453000099', '0453000099', '0453000099']
+        new_partners, new_records = cls._get_sms_test_records(mobile_numbers=mobile_numbers, count=10)
+        records += new_records
+        partners += new_partners
```
Comment (std-odoo) [inline addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:34]:
```suggestion
        cls.partners += new_partners

**Review by std-odoo — COMMENTED** [review]
@datonefaridze Hi,

Some last minutes requests, can you also update the first commit message and explaining why we add a test ?

Thanks

File: addons/test_mass_mailing/tests/test_mailing_statistics_sms.py [hunk 14]
```diff
@@ -24,6 +24,22 @@ def setUpClass(cls):
...
+
+        cls.records = cls._reset_mail_context(records)
```
[the remark this answers is not in the document]
Comment (datonefaridze) [reply addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:32]: 1 sec i pushed wrong commt

[the remark this answers is not in the document]
Comment (datonefaridze) [reply addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:30]: @std-odoo can you check it one more time?

File: addons/test_mass_mailing/tests/test_mailing_statistics_sms.py [hunk 15]
```diff
@@ -43,13 +48,23 @@ def test_mailing_statistics_sms(self):
         for record_idx in (0, 2, 3):
             self.gateway_sms_click(mailing, target_records[record_idx])
 
-        # check mailing statistics
+        for record_idx in [7, 8]:
```
Comment (std-odoo) [inline addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:51]:
```suggestion
        for record_idx in (7, 8):

File: addons/test_mass_mailing/tests/common.py [hunk 16]
```diff
@@ -138,3 +138,21 @@ def setUpClass(cls):
...
+            'country_id': country_be_id,
+            'mobile': mobile_numbers[x]
+        } for x, mobile_number in enumerate(mobile_numbers)])
+        records = cls.env['mail.test.sms'].with_context(**cls._test_context).create([{
+            'name': f'MassSMSTest_{x}',
+            'customer_id': partner.id,
+            'phone_nbr': mobile_number
+        } for x, (mobile_number, partner) in enumerate(zip(mobile_numbers, partners))])
+
+        return partners, records
```
Comment (std-odoo) [inline addons/test_mass_mailing/tests/common.py:158]: now that we don't use Partner, we can just return records
```suggestion
        return records

File: addons/test_mass_mailing/tests/test_mailing_statistics_sms.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:30]: which test fail without `_reset_mail_context` ?

File: addons/test_mass_mailing/tests/test_mailing_statistics_sms.py [hunk 17]
```diff
@@ -24,6 +23,12 @@ def setUpClass(cls):
             name='Marie Marketing',
             signature='--\nMarie'
         )
+        records = cls.records
+        mobile_numbers = [f'045300{x}{x}99' for x in range(6)] + ['0453000099'] * 4
+        _, new_records = cls._get_sms_test_records(mobile_numbers=mobile_numbers)
+        records += new_records
+        cls.records = cls._reset_mail_context(records)
```
Comment (std-odoo) [inline addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:30]: you don't need a variable anymore
either of of them, depending if one test fail or not
```py
cls.records += cls._get_sms_test_records(mobile_numbers=mobile_numbers)
cls.records += cls._reset_mail_context(cls._get_sms_test_records(mobile_numbers=mobile_numbers))
```

**Review by std-odoo — COMMENTED** [review]
@datonefaridze Hi :) Almost good :)

File: addons/test_mass_mailing/tests/test_mailing_statistics_sms.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (datonefaridze) [reply addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:30]: @std-odoo sent one, because i think it doesn't send sms es. i will check comments later i started working on another PR

[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/test_mass_mailing/tests/test_mailing_statistics_sms.py]: > haven't investigated that yet

just ping me once when you are done with everything please ^^

**Comment by datonefaridze** [comment]
@std-odoo i am done, i think everythig should be fine

File: addons/test_mass_mailing/tests/test_mailing_statistics_sms.py [hunk 18]
```diff
@@ -24,6 +23,9 @@ def setUpClass(cls):
             name='Marie Marketing',
             signature='--\nMarie'
         )
+        mobile_numbers = [f'045300{x}{x}99' for x in range(6)] + ['0453000099'] * 4
+        cls.records += cls._get_sms_test_records(mobile_numbers=mobile_numbers)
+        cls.records += cls._reset_mail_context(cls.records)
```
Comment (std-odoo) [inline addons/test_mass_mailing/tests/test_mailing_statistics_sms.py:28]: wut ?
```suggestion
        cls.records = cls._reset_mail_context(cls.records)

**Review by std-odoo — COMMENTED** [review]
@datonefaridze One last :)

File: addons/test_mass_mailing/tests/test_mailing_statistics_sms.py (hunk 18 above)
Comment (datonefaridze) [reply addons/test_mass_mailing/tests/test_mailing_statistics_sms.py]: @std-odoo done

**Review by std-odoo — APPROVED** [review]
LGTM, thanks @datonefaridze

**Comment by std-odoo** [comment]
@robodoo r+
